### PR TITLE
Add /v1/users/login

### PR DESCRIPTION
**Pull request by HinataIzayoi**
Add /v1/users/login to allow users to login with valid email and password combination. 
If user is already logged in, create a new token. If the email is not registered or the email and password combination is not valid, reject login. If success, generate and save token, return apiSuccess with generated token. 

Add service/tokenService to handle creating token, saving tokens, checking if token is valid and/or in use, as well as getting userId with token. 

Add /v1/utils, middleware constructor to handle type checking

#5

### CODE REVIEW

File: .eslintrc.json [hunk 1]
```diff
@@ -1,20 +1,20 @@
...
-   "rules": {
-    "linebreak-style": "off",
-   "require-jsdoc": "off",
-   "max-len": "off",
-   "prefer-spread": "off"
-   }
- }
- 
\ No newline at end of file
+	"extends": ["eslint:recommended", "google"],
```
Comment (Kailang) [inline .eslintrc.json]: If you push master now, this change should be overwritten.

File: controllers/userController.js [hunk 2]
```diff
@@ -23,3 +24,24 @@ exports.register = async function(params) {
...
+
+
+exports.login = async function(params) {
+  const user = await User.findOne({email: params.email});
+  if (!user) {
+    return apiError('Invalid email');
+  }
+  // if user is logged in, reject login
+  if (checkLoggedIn(user.id)) {
+    return apiError('Already logged in');
```
Comment (Kailang) [inline controllers/userController.js]: If logged in, invalidate old token and generate a new token.

File: controllers/userController.js [hunk 3]
```diff
@@ -23,3 +24,24 @@ exports.register = async function(params) {
...
+  }
+
+  const hash = calcPasswordHash(params.password, user.passwordSalt);
+  if (hash !== user.passwordSha256 ) {
+    return apiError('Fail');
+  } else {
```
Comment (Kailang) [inline controllers/userController.js]: Unnecessary else block.

File: controllers/utils.js [hunk 4]
```diff
@@ -19,6 +19,12 @@ exports.genPasswordSalt = function() {
   return crypto.randomBytes(PASSWORD_SALT_LENGTH).toString(PASSWORD_ENCODING);
 };
 
+exports.genSecureRandomString = function(length) {
```
Comment (Kailang) [inline controllers/utils.js]: Refactor genPasswordSalt().

File: services/tokenService.js [hunk 5]
```diff
@@ -0,0 +1,43 @@
+const tokensToId = {};
+const IdToToken = {};
+
+const {genSecureRandomString} = require('../controllers/utils');
+
+
```
Comment (Kailang) [inline services/tokenService.js]: Extra new line.

File: services/tokenService.js [hunk 6]
```diff
@@ -0,0 +1,43 @@
+const tokensToId = {};
+const IdToToken = {};
```
Comment (Kailang) [inline services/tokenService.js]: Dictionary naming:
userIdByToken;
tokenByUserId;

File: services/tokenService.js [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+exports.createToken = function(userId) {
+  const token = genSecureRandomString();
+  saveToken(token, userId);
+  return token;
+};
+
+
+exports.checkTokenValid = function(token) {
+  const userId = checkToken(token);
```
Comment (Kailang) [inline services/tokenService.js]: Inline function call.

File: services/tokenService.js [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+function saveToken(token, id) {
+  IdToToken[id] = token;
+  tokensToId[token] = id;
+}
+
+
+exports.createToken = function(userId) {
+  const token = genSecureRandomString();
+  saveToken(token, userId);
```
Comment (Kailang) [inline services/tokenService.js]: Inline function call.

File: services/tokenService.js [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+  return token;
+};
+
+
+exports.checkTokenValid = function(token) {
+  const userId = checkToken(token);
+  if (!userId) {
+    return false;
+  } else return true;
```
Comment (Kailang) [inline services/tokenService.js]: Remove else.

File: services/tokenService.js [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+};
+
+
+exports.getUserId = function(token) {
+  console.log(tokensToId);
```
Comment (Kailang) [inline services/tokenService.js]: Remove console log.

File: services/tokenService.js [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+  return tokensToId[token];
+};
+
+
+exports.checkLoggedIn = function(userId) {
```
Comment (Kailang) [inline services/tokenService.js]: checkUserHasToken()

File: .eslintrc.json (hunk 1 above)
Comment (Kailang) [reply .eslintrc.json]: Done.

File: controllers/userController.js (hunk 2 above)
Comment (HinataIzayoi) [reply controllers/userController.js]: Done.

File: controllers/userController.js (hunk 3 above)
Comment (HinataIzayoi) [reply controllers/userController.js]: Acknowledged.

File: controllers/utils.js (hunk 4 above)
Comment (HinataIzayoi) [reply controllers/utils.js]: Done.

File: services/tokenService.js (hunk 5 above)
Comment (HinataIzayoi) [reply services/tokenService.js]: Done.

File: services/tokenService.js (hunk 6 above)
Comment (HinataIzayoi) [reply services/tokenService.js]: Done.

File: services/tokenService.js (hunk 7 above)
Comment (HinataIzayoi) [reply services/tokenService.js]: Done.

File: services/tokenService.js (hunk 8 above)
Comment (HinataIzayoi) [reply services/tokenService.js]: Done.

File: services/tokenService.js (hunk 9 above)
Comment (HinataIzayoi) [reply services/tokenService.js]: Done.

File: services/tokenService.js (hunk 10 above)
Comment (HinataIzayoi) [reply services/tokenService.js]: Done.

File: services/tokenService.js (hunk 11 above)
Comment (HinataIzayoi) [reply services/tokenService.js]: Done.

File: controllers/userController.js [hunk 12]
```diff
@@ -23,3 +24,23 @@ exports.register = async function(params) {
 
   return apiSuccess();
 };
+
+
```
Comment (Kailang) [inline controllers/userController.js]: Remove extra newline.

File: controllers/userController.js [hunk 13]
```diff
@@ -23,3 +24,23 @@ exports.register = async function(params) {
...
+exports.login = async function(params) {
+  const user = await User.findOne({email: params.email});
+  if (!user) {
+    return apiError('Invalid email');
+  }
+
+  if (checkUserHasToken(user.id)) {
```
Comment (Kailang) [inline controllers/userController.js]: Remove useless logic.

File: routes/v1/usersRouter.js [hunk 14]
```diff
@@ -1,16 +1,37 @@
...
-  const userName = String(req.body.userName);
-  const email = String(req.body.email);
-  const displayName = String(req.body.displayName);
-  const password = String(req.body.password);
+router.post('/register', createTypeChecker({
+  'userName': STRING,
+  'email': STRING,
+  'displayName': STRING,
+  'password': STRING,
+
```
Comment (Kailang) [inline routes/v1/usersRouter.js]: Remove newline.

File: routes/v1/utils.js [hunk 15]
```diff
@@ -0,0 +1,28 @@
+const {apiError} = require('../../controllers/utils');
+
+exports.STRING = 'string';
+exports.NUMBER = 'number';
+exports.DATE = 'date';
+
```
Comment (Kailang) [inline routes/v1/utils.js]: Remove newline.

File: routes/v1/utils.js [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+
+exports.createTypeChecker = function(params) {
+  return function(req, res, next) {
+    for (const key in params) {
+      if (key[0] === '-' && req.body[key] === undefined) {
+        continue;
+      }
+      if (!checkType(params[key], req.body[key])) {
```
Comment (Kailang) [inline routes/v1/utils.js]: What about optional field?

File: routes/v1/utils.js [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+
+exports.STRING = 'string';
+exports.NUMBER = 'number';
+exports.DATE = 'date';
+
+
+exports.createTypeChecker = function(params) {
+  return function(req, res, next) {
+    for (const key in params) {
+      if (key[0] === '-' && req.body[key] === undefined) {
```
Comment (Kailang) [inline routes/v1/utils.js]: Confusion logic. Split to "required" and "optional" cases.

File: routes/v1/utils.js [hunk 18]
```diff
@@ -0,0 +1,28 @@
...
+        continue;
+      }
+      if (!checkType(params[key], req.body[key])) {
+        return res.json(apiError('Fail'));
```
Comment (Kailang) [inline routes/v1/utils.js]: 'Fail' is not good.
Add some error codes as constants to "controllers/utils.js"
See https://www.restapitutorial.com/httpstatuscodes.html

File: routes/v1/utils.js [hunk 19]
```diff
@@ -0,0 +1,28 @@
...
+      }
+    }
+    next();
+  };
+};
+
+
+function checkType(type, variable) {
```
Comment (Kailang) [inline routes/v1/utils.js]: Inline this function.

File: services/tokenService.js [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+  tokenByUserId[userId] = token;
+  userIdByToken[token] = userId;
+
+  return token;
+};
+
+
+exports.checkTokenValid = function(token) {
+  const userId = userIdByToken[token];
+  if (!userId) {
```
Comment (Kailang) [inline services/tokenService.js]: Add an inline comment.
```
if (!userId) {  // User does not have token.
```

File: controllers/userController.js (hunk 12 above)
Comment (HinataIzayoi) [reply controllers/userController.js]: Done.

File: controllers/userController.js (hunk 13 above)
Comment (HinataIzayoi) [reply controllers/userController.js]: Done.

File: routes/v1/usersRouter.js (hunk 14 above)
Comment (HinataIzayoi) [reply routes/v1/usersRouter.js]: Done.

File: routes/v1/utils.js (hunk 15 above)
Comment (HinataIzayoi) [reply routes/v1/utils.js]: Done.

File: routes/v1/utils.js (hunk 16 above)
Comment (HinataIzayoi) [reply routes/v1/utils.js]: Acknowledged.

File: routes/v1/utils.js (hunk 17 above)
Comment (HinataIzayoi) [reply routes/v1/utils.js]: Acknowledged.

File: routes/v1/utils.js (hunk 19 above)
Comment (HinataIzayoi) [reply routes/v1/utils.js]: Done.

File: routes/v1/utils.js (hunk 18 above)
Comment (HinataIzayoi) [reply routes/v1/utils.js]: Acknowledged.

File: services/tokenService.js (hunk 20 above)
Comment (HinataIzayoi) [reply services/tokenService.js]: Done.

File: routes/v1/utils.js (hunk 16 above)
Comment (HinataIzayoi) [reply routes/v1/utils.js]: Acknowledged.

**Pull request merged**